File: animations-java/src/main/java/dev/hogoshi/animations/animation/AbstractAnimation.java
package dev.hogoshi.animations.animation;

import dev.hogoshi.animations.easing.Easing;
import lombok.Data;

/**
 * Animation basic abstraction
 *
 * @author hogoshi
 */
@Data
public abstract class AbstractAnimation {

    private double value;
    private long start;
    private double duration;
    private double from;
    private double to;
    private Easing easing;
    private boolean debug;

    /**
     * Use it before rendering element which uses animation's value
     *
     * @return boolean which means if animation is alive
     */
    public boolean update() {
        boolean alive = isAlive();

        if (alive) {
            this.value = interpolate(getFrom(), getTo(), getEasing().ease(calculatePart()));
        } else {
            setStart(0);
            this.value = getTo();
        }

        return alive;
    }

    /**
     * @return boolean which means if animation is alive
     */
    public boolean isAlive() {
        return !isDone();
    }

    /**
     * @return boolean which means if animation is done
     */
    public boolean isDone() {
        return calculatePart() >= 1.0;
    }

    /**
     * @return calculates current x from 0 to 1 for calculating easing's current value
     */
    public double calculatePart() {
        return (double) (System.currentTimeMillis() - getStart()) / getDuration();
    }

    /**
     * @return linear interpolation
     */
    public double interpolate(double start, double end, double pct) {
        return start + (end - start) * pct;
    }

}

File: animations-kotlin/src/main/kotlin/dev/hogoshi/animations/animation/AbstractAnimation.kt
package dev.hogoshi.animations.animation

import dev.hogoshi.animations.easing.Easings

/**
 * Animation basic abstraction
 * @author hogoshi
 */
abstract class AbstractAnimation {

    var value = 0.0
    var start = 0L
    var duration = 0.0
    var from = 0.0
    var to = 0.0
    var easing = Easings.LINEAR
    var debug = false

    /**
     * Use it before rendering element which uses animation's value
     * @return boolean which means if animation is alive
     */
    fun update(): Boolean {
        val alive = isAlive()

        if (alive) {
            value = interpolate(from, to, easing.ease(calculatePart()))
        } else {
            start = 0
            value = to
        }

        return alive
    }

    /**
     * @return boolean which means if animation is alive
     */
    fun isAlive() = !isDone()

    /**
     * @return boolean which means if animation is done
     */
    fun isDone() = calculatePart() >= 1.0

    /**
     * @return calculates current x from 0 to 1 for calculating easing's current value
     */
    private fun calculatePart() = (System.currentTimeMillis() - start).toDouble() / duration

    /**
     * @return linear interpolation
     */
    private fun interpolate(start: Double, end: Double, pct: Double) = start + (end - start) * pct

}